Framework/DataHandling/test/SaveGDATest.h

```diff
@@ -0,0 +1,308 @@
+    testAlg.initialize();
+    TS_ASSERT_THROWS(testAlg.setProperty("InputWorkspace", "ws"),
+                     std::invalid_argument);
+
+    ADS.remove("ws");
+  }
+
+  void test_groupingSchemeMustMatchNumberOfSpectra() {
+    SaveGDA testAlg;
+    testAlg.initialize();
+    testAlg.setProperty("InputWorkspace", INPUT_GROUP_NAME);
+    // This should make the algorithm throw, as there are 2 spectra but three
+    // values in the grouping scheme
+    testAlg.setProperty("GroupingScheme", std::vector<int>{1, 2, 3});
+    testAlg.setProperty("GSASParamFile", m_paramsFile.path());
+
+    Poco::TemporaryFile tempFile;
+    const std::string &tempFileName = tempFile.path();
+    TS_ASSERT_THROWS_NOTHING(testAlg.setProperty("Filename", tempFileName));
+
+    TS_ASSERT_THROWS_ANYTHING(testAlg.execute());
+  }
+
+  void test_algExecutesWithValidInput() {
+    SaveGDA testAlg;
+
+    testAlg.initialize();
+    TS_ASSERT_THROWS_NOTHING(
+        testAlg.setProperty("InputWorkspace", INPUT_GROUP_NAME));
+    TS_ASSERT_THROWS_NOTHING(
+        testAlg.setProperty("GSASParamFile", m_paramsFile.path()));
+    TS_ASSERT_THROWS_NOTHING(
+        testAlg.setProperty("GroupingScheme", std::vector<int>{1, 2}));
+
+    Poco::TemporaryFile tempFile;
+    const std::string &tempFileName = tempFile.path();
+    TS_ASSERT_THROWS_NOTHING(testAlg.setProperty("Filename", tempFileName));
+
+    TS_ASSERT_THROWS_NOTHING(testAlg.execute());
+    TS_ASSERT(testAlg.isExecuted());
+
+    Poco::File shouldExist(tempFileName);
+    TS_ASSERT(shouldExist.exists());
+  }
+
+  void test_headerValuesAreCorrect() {
+    SaveGDA testAlg;
+    testAlg.initialize();
+    testAlg.setProperty("InputWorkspace", INPUT_GROUP_NAME);
+    testAlg.setProperty("GSASParamFile", m_paramsFile.path());
+    testAlg.setProperty("GroupingScheme", std::vector<int>({2, 3}));
+    Poco::TemporaryFile tempFile;
+    const std::string &tempFileName = tempFile.path();
+    testAlg.setProperty("Filename", tempFileName);
+    testAlg.execute();
+
+    std::ifstream file(tempFileName);
+    std::string line;
+    TS_ASSERT(file.is_open());
+
+    // first line is header
+    std::getline(file, line);
+    std::vector<std::string> headerItems;
+    boost::split(headerItems, line, boost::is_any_of(" "),
+                 boost::token_compress_on);
+
+    int numPoints = 0;
+    int numLines = 0;
+    std::vector<double> TOFs;
+    while (std::getline(file, line) && !boost::starts_with(line, "BANK")) {
+      std::vector<std::string> lineItems;
+      boost::trim(line);
+      boost::split(lineItems, line, boost::is_any_of(" "),
+                   boost::token_compress_on);
+
+      // each point has 3 space-separated items on a line
+      numPoints += static_cast<int>(lineItems.size()) / 3;
+      numLines++;
+
+      for (size_t i = 0; i < lineItems.size(); i += 3) {
+        TOFs.emplace_back(std::stod(lineItems[i]));
+      }
+    }
+
+    const auto expectedNumPoints = std::stoi(headerItems[2]);
```

Comment:
Check the length of the array before accessing into it. An assert failing is far easy to diagnose than an exception violation in a unit test